src/Npgsql/Internal/NpgsqlConnector.cs

```diff
@@ -839,10 +841,10 @@ async Task RawOpen(SslMode sslMode, NpgsqlTimeout timeout, bool async, Cancellat
                         certificateValidationCallback = SslTrustServerValidation;
                         checkCertificateRevocation = false;
                     }
-                    else if ((Settings.RootCertificate ?? PostgresEnvironment.SslCertRoot ?? PostgresEnvironment.SslCertRootDefault) is
-                             { } certRootPath)
+                    else if ((caCert = DataSource.CACertificateCallback?.Invoke()) is not null || (certRootPath = Settings.RootCertificate ?? PostgresEnvironment.SslCertRoot ?? PostgresEnvironment.SslCertRootDefault) is
```

Comment:
Should we have validation against both providing a root certificate and UserCertificateValidationCallback? Because the former just gets ignored if the latter is set...

We could ask the same question about Settings.RootCertificate too, btw.